Browser_IDE/executionEnvironment_Page.js

```diff
@@ -80,34 +91,52 @@ document.getElementById("canvas").addEventListener("click", async function () {
 
 // Convenience function for reporting errors, printing them to the terminal
 // and also sending a message to the main window.
-function ReportError(block, message, line, formatted=false){
+function ReportError(block, message, line, stacktrace ,formatted=false){
     let outputMessage = message != "";
+    let stackTrace = stacktrace;
+
+    // Ensure block and message are strings
+    block = block || "";
+    message = message || "";
 
-    if (outputMessage && line != null)
-        message = "Error on line "+line+": "+message;
+    // Escape only the user-provided input
+    let escapedBlock = escapeHtml(block);
+    let escapedMessage = escapeHtml(message);
 
-    if (block != null && block != "" && block != "__USERCODE__null") {
-        if (!block.startsWith(userCodeBlockIdentifier)){
-            message = "Please file a bug report and send us the following info!\n    Error in file: "+block+"\n    "+message;
-            block = "Internal Error";
+    if (outputMessage && line != null && !formatted)
+        escapedMessage = "Error on line "+line+": "+escapedMessage;
+
+    if (escapedBlock != null && escapedBlock != "" && escapedBlock != "__USERCODE__null") {
+        if (!escapedBlock.startsWith(userCodeBlockIdentifier)){
+            escapedMessage = "Please file a bug report and send us the following info!\n    Error in file: "+escapedBlock+"\n    "+escapedMessage;
+            escapedBlock = "Internal Error";
         }
         else{
-            block = block.slice(userCodeBlockIdentifier.length);
+            escapedBlock = escapedBlock.slice(userCodeBlockIdentifier.length);
         }
         if (outputMessage)
-            message = "(" + block + ") " + message;
+            escapedMessage = "(" + escapedBlock + ") " + escapedMessage;
+    }
+    
+    // Check if the stackTrace is empty
+    if (stackTrace.trim() !== "") {

```

Comment:
Make sure to check `stackTrace != null` first here, since `null` is also passed in by a few functions and without it this throws an exception (which unfortunately gets caught and silenced, hence why no errors show up - totally understand that you didn't catch this, I almost missed it too! 🤣 ).